### ISSUE TITLE
Cannot have epoch in version

### ISSUE THREAD

**Issue by noscript** [openSUSE/obs-build#465]
Trying to set epoch in `_service` file like so
```xml
<param name="versionrewrite-replacement">2:\1</param>
```
gives me
```
Stripped epoch from Version field, which is now "8.1.0369.git0+12.1".
```

I am aware of downsides of using epoch but I in personal OBS repositories there shouldn't be such limitation.

**Comment by mlschroe** [comment]
But isn't the version in debtransform only used to generate filenames? I.e. isn't the version that goes into to built deb files taken from the changelog and thus epochs are working? Does the tar_scm service even understand debian builds?

But maybe I didn't understand your issue report.

**Comment by noscript** [comment]
> isn't the version that goes into to built deb files taken from the changelog 

Before the build, my [changelog](https://build.opensuse.org/package/view_file/home:sergey-vlasov/vim/_service:set_version:tar_scm:debian.changelog?expand=1) contains epoch (thanks to `set_version`) but then it is not present in artifacts' changelog.

>Does the tar_scm service even understand debian builds?

Why should it. At least it creates tarballs with correct filename (with epoch).

>But isn't the version in debtransform only used to generate filenames?

The generated version strings affects everything. As the result, epoch disappears from all [artifacts](https://build.opensuse.org/package/binaries/home:sergey-vlasov/vim/Debian_9.0).

Please see my attempt to package [vim](https://build.opensuse.org/package/show/home:sergey-vlasov/vim).

**Comment by casusbelli** [comment]
Hitting the same spot, did somebody find a solution/workaround for this?

**Comment by noscript** [comment]
I had to set to my OBS repository a higher priority in Debian. However I would definitely prefer using epoch instead.

**Comment by m-ildefons** [comment]
This is broken for me too. I think this commit is the problem: https://github.com/openSUSE/obs-build/commit/81017d36161c563a62bb8cf0482759510d3748b5
It just flat out removes the epoch from the version.

@abitrolly : How is epoch handling supposed to work with this?
Since 
https://github.com/openSUSE/obs-build/commit/f4cf063683fc159a6778507f5ce8418c4f5f2bc2, the epoch is removed from the version string. This means saving back the modified version also removes the epoch, which is too much.

**Comment by noscript** [comment]
Issue was opened 3 years ago. How could 81017d3 be the problem?

**Comment by m-ildefons** [comment]
You are right, I think the origin of the problem is https://github.com/openSUSE/obs-build/commit/f4cf063683fc159a6778507f5ce8418c4f5f2bc2 . My observation also was not made with a `_service` file but rather a `.dsc` file directly. But its still the same symptom - epochs vanishing from the version string.

After reading the code, as far as I understand it does this:
- read the dsc file
- remove epoch (introduced by https://github.com/openSUSE/obs-build/commit/f4cf063683fc159a6778507f5ce8418c4f5f2bc2)
- add debian revision
- add debtransform release
- save version string back to the map of things to write out into the dsc file (introduced by https://github.com/openSUSE/obs-build/commit/81017d36161c563a62bb8cf0482759510d3748b5)
- remove debian revision and everything after from version
- write out files

I believe moving the removal of the epoch down after the save-back-to-map step should fix this.

**Comment by abitrolly** [comment]
@noscript @m-ildefons could be right, epoch was always removed since https://github.com/openSUSE/obs-build/commit/c4737b6bc7f5e0893609bd0fbcb7d1af6d3a4d6d but transformed version was not written back to `.dsc` until https://github.com/openSUSE/obs-build/commit/53e20ebf890c7c140e7846a3ce82078d0ea79fef where it was written only when `DEBTRANSFORM-RELEASE` is introduced.

In https://github.com/openSUSE/obs-build/commit/81017d36161c563a62bb8cf0482759510d3748b5 I just made all changes to version to be recorded unconditionally as the modified version is used in many places, and epoch version is not used. It kind of strange to use `DEBTRANSFORM-RELEASE` and have you epoch removed, so I thought it is a bug that it is not always removed when there is a specific "no epoch in version, please" comment.

**Comment by abitrolly** [comment]
I wouldn't mind leaving the epoch alone if somebody added it to `.dsc`. However, I feel the lack of total understanding of why would anyone want epoch or ask for no epoch, please. I wish `debtransform` had more tests explaining this - waiting for them to be merged in CI - https://github.com/openSUSE/obs-build/pull/650

@mlschroe do remember why you added this version?

https://github.com/openSUSE/obs-build/commit/c4737b6bc7f5e0893609bd0fbcb7d1af6d3a4d6d#diff-f985383c757e4b0e7eaafe8f84ca80e9d1fb3939a82210e593de49bd070ab5d4R213

**Comment by noscript** [comment]
There are reasons not to want epoch, which makes sense for distributions. Once you ship a package with epoch you get stuck with it forever. Because epoch has the highest priority in version number. Meaning that users who already have some package installed with epoch will only get updates of that package if it has same or higher epoch.

But as I said in the original post, it should not matter for personal OBS repositories.

**Comment by abitrolly** [comment]
I would inspect some previous archive produced by OBS with initial epoch version. Even if resulting .dsc file contains the epoch number, no other filename or file content should contain it if I read the code right. I wonder what happened when Debian compiled that source package into binary and then user installed the resulting archive. Which version number it is in the end.

**Comment by m-ildefons** [comment]
> There are reasons not to want epoch, which makes sense for distributions. Once you ship a package with epoch you get stuck with it forever. Because epoch has the highest priority in version number. Meaning that users who already have some package installed with epoch will only get updates of that package if it has same or higher epoch.
> 
> But as I said in the original post, it should not matter for personal OBS repositories.

There are good reasons to keep epoch as well. As you said, once a package is shipped with it, you can not easily reverse it. This is a problem for self hosted OBS that is used e.g. for commercial software. Just dropping it in that case creates unreasonable complications for release. For me this essentially means not being able to have the epoch is a major pain.

As far as my understanding for the epoch number goes, it is meant to be a tool to deal with version number schema changes upstream of a distribution. 

Unfortunately, neither documentation nor wiki really show how this is supposed to work in OBS.
https://en.opensuse.org/openSUSE:Build_Service_Debian_builds
https://openbuildservice.org/help/manuals/obs-user-guide/cha.obs.package_formats.html#id-1.5.6.2.7
